From e2875e9d1861658cb3b7af83a59538929dded184 Mon Sep 17 00:00:00 2001
From: Victor Bo <bvoid@yandex.ru>
Date: Wed, 11 Mar 2020 20:20:05 +0200
Subject: [PATCH 13/24] add three-fingers-swipe to screenshot [2/2]

  * based on
  https://github.com/Evolution-X/packages_apps_Evolver/commit/1b64fadf5debcda66382763312d77867bbb8be78
  https://github.com/Evolution-X/packages_apps_Evolver/blob/7f3cfe97bf3d98649f9b4c123e5bd24e6665953d/src/com/evolution/settings/preference/SystemSettingSwitchPreference.java

	modified:   res/values/strings.xml
	modified:   res/xml/gestures.xml

	modified:   res/values/strings.xml
	modified:   res/xml/gestures.xml
	new file:   src/com/caos/settings/preference/SystemSettingSwitchPreference.java
---
 res/values/strings.xml                        |  5 ++
 res/xml/gestures.xml                          |  6 ++
 .../SystemSettingSwitchPreference.java        | 64 +++++++++++++++++++
 3 files changed, 75 insertions(+)
 create mode 100644 src/com/caos/settings/preference/SystemSettingSwitchPreference.java

diff --git a/res/values/strings.xml b/res/values/strings.xml
index 42e4b05451..0de8cc089d 100644
--- a/res/values/strings.xml
+++ b/res/values/strings.xml
@@ -12178,4 +12178,9 @@
     <string name="media_controls_no_players">No players available</string>
     <!-- Keywords for the media controls setting [CHAR LIMIT=NONE]-->
     <string name="keywords_media_controls">media</string>
+
+    <!-- Three-fingers-swipe to screenshot -->
+    <string name="three_finger_gesture">Swipe to screenshot</string>
+    <string name="three_finger_gesture_summary">Swipe down with three fingers to take a screenshot</string>
+
 </resources>
diff --git a/res/xml/gestures.xml b/res/xml/gestures.xml
index 0f407f4363..c292e40bcc 100644
--- a/res/xml/gestures.xml
+++ b/res/xml/gestures.xml
@@ -86,4 +86,10 @@
         android:title="@string/power_menu_setting_name"
         android:fragment="com.android.settings.gestures.PowerMenuSettings"
         settings:controller="com.android.settings.gestures.PowerMenuPreferenceController" />
+
+    <com.caos.settings.preference.SystemSettingSwitchPreference
+        android:key="three_finger_gesture"
+        android:title="@string/three_finger_gesture"
+        android:summary="@string/three_finger_gesture_summary"
+        android:defaultValue="false" />
 </PreferenceScreen>
diff --git a/src/com/caos/settings/preference/SystemSettingSwitchPreference.java b/src/com/caos/settings/preference/SystemSettingSwitchPreference.java
new file mode 100644
index 0000000000..0e1a4a9bde
--- /dev/null
+++ b/src/com/caos/settings/preference/SystemSettingSwitchPreference.java
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2019-2020 The Evolution X Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.caos.settings.preference;
+
+import android.content.Context;
+import android.provider.Settings;
+import androidx.preference.SwitchPreference;
+import android.util.AttributeSet;
+
+public class SystemSettingSwitchPreference extends SwitchPreference {
+    public SystemSettingSwitchPreference(Context context, AttributeSet attrs, int defStyle) {
+        super(context, attrs, defStyle);
+    }
+
+    public SystemSettingSwitchPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    public SystemSettingSwitchPreference(Context context) {
+        super(context, null);
+    }
+
+    @Override
+    protected boolean persistBoolean(boolean value) {
+        if (shouldPersist()) {
+            if (value == getPersistedBoolean(!value)) {
+                // It's already there, so the same as persisting
+                return true;
+            }
+            Settings.System.putInt(getContext().getContentResolver(), getKey(), value ? 1 : 0);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    protected boolean getPersistedBoolean(boolean defaultReturnValue) {
+        if (!shouldPersist()) {
+            return defaultReturnValue;
+        }
+        return Settings.System.getInt(getContext().getContentResolver(),
+                getKey(), defaultReturnValue ? 1 : 0) != 0;
+    }
+
+    @Override
+    protected void onSetInitialValue(boolean restoreValue, Object defaultValue) {
+        setChecked(Settings.System.getString(getContext().getContentResolver(), getKey()) != null ? getPersistedBoolean(isChecked())
+                : (Boolean) defaultValue);
+    }
+}
-- 
2.17.1

